tests/integrations/test_middleware_factory.py

```diff
@@ -0,0 +1,414 @@
+    base_stream = BaseStream(
+        stream=(t for t in stream_chunks),
+        metadata={},
+        tool_types=[],
+        call_response_type=MagicMock,
+        model="model",
+        prompt_template="prompt_template",
+        fn_args={},
+        dynamic_config=None,
+        messages=[],
+        call_params={},
+        call_kwargs={},
+    )  # type: ignore
+    my_structured_stream = BaseStructuredStream(stream=base_stream, response_model=Foo)
+
+    def sync_fn() -> BaseStructuredStream:
+        return my_structured_stream
+
+    def handle_structured_stream(result, fn, context):
+        assert isinstance(result, BaseStructuredStream)
+        if context:
+            assert isinstance(context, str)
+
+    @contextmanager
+    def my_context_manager(fn):
+        yield "foo"
+
+    decorate = middleware_decorator(
+        sync_fn, handle_structured_stream=handle_structured_stream
+    )
+    result = decorate()
+    for chunk in result:
+        assert chunk.model_dump() == my_foo.model_dump()
+    base_stream.stream = (t for t in stream_chunks)  # Re-add the iterator
+    decorate_with_context = middleware_decorator(
+        sync_fn,
+        handle_structured_stream=handle_structured_stream,
+        custom_context_manager=my_context_manager,
+    )
+    for chunk, _ in decorate_with_context():
+        pass
+
+
+@pytest.mark.asyncio
+async def test_middleware_decorator_structured_stream_async():
+    patch.multiple(BaseStream, __abstractmethods__=set()).start()
+
+    class Foo(BaseModel):
+        bar: str
+        baz: int
+
+    mock_chunk = MagicMock()
+    my_foo = Foo(bar="bar", baz=1)
+    mock_chunk.content = my_foo.model_dump_json()
+    mock_chunk.input_tokens = 1
+    mock_chunk.output_tokens = 2
+    mock_chunk.model = "updated_model"
+
+    base_stream = MagicMock()
+    base_stream.__iter__.return_value = (t for t in [(mock_chunk, None)])
+
+    async def generator(self):
+        yield mock_chunk, None
+
+    base_stream.__aiter__ = generator
+    my_structured_stream = BaseStructuredStream(stream=base_stream, response_model=Foo)
+
+    async def async_fn() -> BaseStructuredStream:
+        return my_structured_stream
+
+    async def handle_structured_stream_async(result, fn, context):
+        assert isinstance(result, BaseStructuredStream)
+        if context:
+            assert isinstance(context, str)
+
+    @contextmanager
+    def my_context_manager(fn):
+        yield "foo"
+
+    decorate = middleware_decorator(
+        async_fn, handle_structured_stream_async=handle_structured_stream_async
+    )
+    result = await decorate()
+    async for chunk in result:
+        assert chunk.model_dump() == my_foo.model_dump()
+    decorate_with_context = middleware_decorator(
+        async_fn,
+        handle_structured_stream_async=handle_structured_stream_async,
+        custom_context_manager=my_context_manager,
+    )
+    async for chunk, _ in await decorate_with_context():
+        pass
```

Comment:
here and all other tests I'm a bit confused on what we're testing by running the function? Just that it runs?